File: packages/go/core/store/sqlite.go
package store

import (
	"context"
	"database/sql"
	"fmt"
	"os"
	"path/filepath"
	"time"

	_ "modernc.org/sqlite"
)

type SQLiteHitStore struct {
	db *sql.DB
}

func NewSQLiteHitStore(path string) (*SQLiteHitStore, error) {
	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
		return nil, fmt.Errorf("create db directory: %w", err)
	}
	db, err := sql.Open("sqlite", path)
	if err != nil {
		return nil, err
	}
	db.SetMaxOpenConns(1) // SQLite is single-writer
	if err := initSQLiteSchema(db); err != nil {
		db.Close()
		return nil, err
	}
	return &SQLiteHitStore{db: db}, nil
}

func initSQLiteSchema(db *sql.DB) error {
	stmts := []string{
		"PRAGMA journal_mode = WAL",
		`CREATE TABLE IF NOT EXISTS hits (
			ts            TEXT NOT NULL,
			path          TEXT NOT NULL,
			method        TEXT NOT NULL,
			caller_id     TEXT,
			user_agent    TEXT,
			days_left     INTEGER,
			endpoint_key  TEXT NOT NULL,
			enforced      INTEGER NOT NULL DEFAULT 0
		)`,
		"CREATE INDEX IF NOT EXISTS idx_endpoint ON hits(endpoint_key)",
		"CREATE INDEX IF NOT EXISTS idx_ts ON hits(ts)",
	}
	for _, stmt := range stmts {
		if _, err := db.Exec(stmt); err != nil {
			return err
		}
	}
	return nil
}

func (s *SQLiteHitStore) Record(hit HitEvent) error {
	enforced := 0
	if hit.Enforced {
		enforced = 1
	}
	_, err := s.db.ExecContext(context.Background(),
		"INSERT INTO hits VALUES (?,?,?,?,?,?,?,?)",
		hit.Ts.UTC().Format(time.RFC3339Nano),
		hit.Path,
		hit.Method,
		nullStr(hit.CallerID),
		nullStr(hit.UserAgent),
		nullInt(hit.DaysUntilSunset),
		hit.EndpointKey,
		enforced,
	)
	return err
}

func (s *SQLiteHitStore) RecentHits(query HitQuery) ([]HitEvent, error) {
	sinceDays := query.SinceDays
	if sinceDays == 0 {
		sinceDays = 30
	}
	limit := query.Limit
	if limit == 0 {
		limit = 100
	}
	since := time.Now().UTC().AddDate(0, 0, -sinceDays).Format(time.RFC3339Nano)

	q := `SELECT ts, path, method, caller_id, user_agent, days_left, endpoint_key, enforced
	      FROM hits WHERE ts >= ? `
	args := []interface{}{since}

	if query.EndpointKey != nil {
		q += "AND endpoint_key = ? "
		args = append(args, *query.EndpointKey)
	}
	if query.CallerID != nil {
		q += "AND caller_id = ? "
		args = append(args, *query.CallerID)
	}

	q += "ORDER BY ts DESC LIMIT ?"
	args = append(args, limit)

	rows, err := s.db.QueryContext(context.Background(), q, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var hits []HitEvent
	for rows.Next() {
		var (
			tsStr       string
			path        string
			method      string
			callerID    sql.NullString
			userAgent   sql.NullString
			daysLeft    sql.NullInt64
			endpointKey string
			enforced    int
		)
		if err := rows.Scan(&tsStr, &path, &method, &callerID, &userAgent, &daysLeft, &endpointKey, &enforced); err != nil {
			return nil, err
		}
		ts, _ := time.Parse(time.RFC3339Nano, tsStr)
		hit := HitEvent{
			Ts:          ts,
			Path:        path,
			Method:      method,
			EndpointKey: endpointKey,
			Enforced:    enforced == 1,
		}
		if callerID.Valid {
			hit.CallerID = &callerID.String
		}
		if userAgent.Valid {
			hit.UserAgent = &userAgent.String
		}
		if daysLeft.Valid {
			d := int(daysLeft.Int64)
			hit.DaysUntilSunset = &d
		}
		hits = append(hits, hit)
	}
	return hits, rows.Err()
}

func (s *SQLiteHitStore) EndpointSummaries(sinceDays int) ([]EndpointSummary, error) {
	if sinceDays == 0 {
		sinceDays = 30
	}
	since := time.Now().UTC().AddDate(0, 0, -sinceDays).Format(time.RFC3339Nano)
	ctx := context.Background()

	rows, err := s.db.QueryContext(ctx, `
		SELECT endpoint_key, COUNT(*) AS total_hits,
		       COUNT(DISTINCT caller_id) AS unique_callers, MAX(ts) AS last_seen
		FROM hits WHERE ts >= ?
		GROUP BY endpoint_key ORDER BY total_hits DESC
	`, since)
	if err != nil {
		return nil, err
	}

	// Collect all outer rows before closing — nested queries below need the connection.
	type epRow struct {
		epKey         string
		totalHits     int
		uniqueCallers int
		lastSeenStr   sql.NullString
	}
	var collected []epRow
	for rows.Next() {
		var r epRow
		if err := rows.Scan(&r.epKey, &r.totalHits, &r.uniqueCallers, &r.lastSeenStr); err != nil {
			rows.Close()
			return nil, err
		}
		collected = append(collected, r)
	}
	rows.Close()
	if err := rows.Err(); err != nil {
		return nil, err
	}

	var summaries []EndpointSummary
	for _, r := range collected {
		callerRows, err := s.db.QueryContext(ctx, `
			SELECT caller_id, COUNT(*) AS cnt FROM hits
			WHERE endpoint_key = ? AND ts >= ? AND caller_id IS NOT NULL
			GROUP BY caller_id ORDER BY cnt DESC LIMIT 10
		`, r.epKey, since)
		if err != nil {
			return nil, err
		}
		var topCallers []TopCaller
		for callerRows.Next() {
			var tc TopCaller
			if err := callerRows.Scan(&tc.CallerID, &tc.Count); err != nil {
				callerRows.Close()
				return nil, err
			}
			topCallers = append(topCallers, tc)
		}
		callerRows.Close()

		sum := EndpointSummary{
			EndpointKey:   r.epKey,
			TotalHits:     r.totalHits,
			UniqueCallers: r.uniqueCallers,
			TopCallers:    topCallers,
		}
		if r.lastSeenStr.Valid {
			t, _ := time.Parse(time.RFC3339Nano, r.lastSeenStr.String)
			sum.LastSeen = &t
		}
		summaries = append(summaries, sum)
	}
	return summaries, nil
}

func (s *SQLiteHitStore) TotalSummary(sinceDays int) (TotalSummary, error) {
	if sinceDays == 0 {
		sinceDays = 30
	}
	since := time.Now().UTC().AddDate(0, 0, -sinceDays).Format(time.RFC3339Nano)
	ctx := context.Background()

	var total TotalSummary
	if err := s.db.QueryRowContext(ctx, `
		SELECT COUNT(*), COUNT(DISTINCT caller_id), COUNT(DISTINCT endpoint_key)
		FROM hits WHERE ts >= ?
	`, since).Scan(&total.TotalHits, &total.UniqueCallers, &total.EndpointsWithTraffic); err != nil {
		return TotalSummary{}, err
	}

	if err := s.db.QueryRowContext(ctx, `
		SELECT COUNT(DISTINCT endpoint_key)
		FROM hits WHERE ts >= ? AND days_left < 0
	`, since).Scan(&total.PastSunsetWithTraffic); err != nil {
		return TotalSummary{}, err
	}

	return total, nil
}

func (s *SQLiteHitStore) Close() error {
	return s.db.Close()
}

func nullStr(s *string) interface{} {
	if s == nil {
		return nil
	}
	return *s
}

func nullInt(i *int) interface{} {
	if i == nil {
		return nil
	}
	return *i
}

File: packages/javascript/src/core/store/sqlite.ts
import fs from 'fs'
import path from 'path'
import { HitStore, HitEvent, HitQuery, EndpointSummary, TotalSummary } from './base'

// Dynamic import so the package installs without better-sqlite3 if not needed
// eslint-disable-next-line @typescript-eslint/no-require-imports
type BetterSqlite3 = typeof import('better-sqlite3')
type Database = InstanceType<BetterSqlite3>

export class SQLiteHitStore extends HitStore {
  private db: Database | null = null

  constructor(private readonly dbPath: string = '.dusk/hits.db') {
    super()
  }

  private getDb(): Database {
    if (!this.db) {
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const Database = require('better-sqlite3') as BetterSqlite3
      fs.mkdirSync(path.dirname(this.dbPath), { recursive: true })
      const db = new Database(this.dbPath)
      db.pragma('journal_mode = WAL')
      db.exec(`
        CREATE TABLE IF NOT EXISTS hits (
          ts            TEXT NOT NULL,
          path          TEXT NOT NULL,
          method        TEXT NOT NULL,
          caller_id     TEXT,
          user_agent    TEXT,
          days_left     INTEGER,
          endpoint_key  TEXT NOT NULL,
          enforced      INTEGER NOT NULL DEFAULT 0
        );
        CREATE INDEX IF NOT EXISTS idx_endpoint ON hits(endpoint_key);
        CREATE INDEX IF NOT EXISTS idx_ts ON hits(ts);
      `)
      this.db = db
    }
    return this.db!
  }

  async record(hit: HitEvent): Promise<void> {
    this.getDb()
      .prepare('INSERT INTO hits VALUES (?,?,?,?,?,?,?,?)')
      .run(
        hit.ts.toISOString(),
        hit.path,
        hit.method,
        hit.caller_id,
        hit.user_agent,
        hit.days_until_sunset,
        hit.endpoint_key,
        hit.enforced ? 1 : 0,
      )
  }

  async recentHits(query: HitQuery = {}): Promise<HitEvent[]> {
    const { endpoint_key, caller_id, since_days = 30, limit = 100 } = query
    const since = new Date(Date.now() - since_days * 86_400_000).toISOString()

    let sql = `
      SELECT ts, path, method, caller_id, user_agent, days_left, endpoint_key, enforced
      FROM hits
      WHERE ts >= ?
    `
    const params: unknown[] = [since]

    if (endpoint_key) { sql += ' AND endpoint_key = ?'; params.push(endpoint_key) }
    if (caller_id)    { sql += ' AND caller_id = ?';    params.push(caller_id) }

    sql += ' ORDER BY ts DESC LIMIT ?'
    params.push(limit)

    const rows = this.getDb().prepare(sql).all(...params) as Array<Record<string, unknown>>
    return rows.map((r) => ({
      ts: new Date(r.ts as string),
      path: r.path as string,
      method: r.method as string,
      caller_id: (r.caller_id as string | null) ?? null,
      user_agent: (r.user_agent as string | null) ?? null,
      days_until_sunset: r.days_left as number | null,
      endpoint_key: r.endpoint_key as string,
      enforced: (r.enforced as number) === 1,
    }))
  }

  async endpointSummaries(since_days = 30): Promise<EndpointSummary[]> {
    const since = new Date(Date.now() - since_days * 86_400_000).toISOString()
    const db = this.getDb()

    const rows = db.prepare(`
      SELECT endpoint_key,
             COUNT(*)                    AS total_hits,
             COUNT(DISTINCT caller_id)   AS unique_callers,
             MAX(ts)                     AS last_seen
      FROM hits
      WHERE ts >= ?
      GROUP BY endpoint_key
      ORDER BY total_hits DESC
    `).all(since) as Array<Record<string, unknown>>

    return rows.map((row) => {
      const callers = db.prepare(`
        SELECT caller_id, COUNT(*) AS cnt
        FROM hits
        WHERE endpoint_key = ? AND ts >= ? AND caller_id IS NOT NULL
        GROUP BY caller_id
        ORDER BY cnt DESC
        LIMIT 10
      `).all(row.endpoint_key, since) as Array<Record<string, unknown>>

      return {
        endpoint_key: row.endpoint_key as string,
        total_hits: row.total_hits as number,
        unique_callers: row.unique_callers as number,
        last_seen: row.last_seen ? new Date(row.last_seen as string) : null,
        top_callers: callers.map((c) => [c.caller_id as string, c.cnt as number]),
      }
    })
  }

  async totalSummary(since_days = 30): Promise<TotalSummary> {
    const since = new Date(Date.now() - since_days * 86_400_000).toISOString()
    const db = this.getDb()

    const row = db.prepare(`
      SELECT COUNT(*)                      AS total_hits,
             COUNT(DISTINCT caller_id)     AS unique_callers,
             COUNT(DISTINCT endpoint_key)  AS endpoints_with_traffic
      FROM hits WHERE ts >= ?
    `).get(since) as Record<string, number>

    const past = db.prepare(`
      SELECT COUNT(DISTINCT endpoint_key)
      FROM hits WHERE ts >= ? AND days_left < 0
    `).pluck().get(since) as number

    return {
      total_hits: row.total_hits,
      unique_callers: row.unique_callers,
      endpoints_with_traffic: row.endpoints_with_traffic,
      past_sunset_with_traffic: past,
    }
  }

  async close(): Promise<void> {
    if (this.db) {
      this.db.close()
      this.db = null
    }
  }
}
